### PR TITLE
Enhance NVIDIA Inference Test Harness, Parameterize GPU Usage, Update CI, and Upgrade PyTorch to v2.5

### PR DESCRIPTION
**Pull request by mattcjo**
*Issue #, if available:*

*Description of changes:*
NVIDIA Inference Test Harness

   1. Parameterization:
        a. Added flags (`--inferenceMode`, `--gpuRequested`) to `main_test.go` for dynamic control of GPU usage and latency/throughput modes.
        b. Updated bert_inference_test.go and the inference manifest (`bert-inference.yaml`) to accept templated GPU values (`GPUPerNode`), ensuring tests can run on a range of GPU capacities.
    c. Improved Logging & Timings:
        * Logs which node each Pod runs on, captures total job duration, and prints structured [BERT_INFERENCE_METRICS].
        * Incorporates realistic NSP-based dummy data in infer.py for more authentic BERT inference testing.

2. GitHub Workflows

    a. `.github/workflows` updated to accommodate the new nvidia-inference path for image builds, aligning continuous integration with the refactored project structure.

3. Kubetest2 Dockerfile

    a. `Dockerfile.kubetest2` modified to include the newly built Go test binaries for the nvidia-inference tests, allowing seamless execution within the Kubetest2 container environment.

4. PyTorch Upgrade

    PyTorch version bumped to`2.5` in relevant files, ensuring better performance and access to up-to-date features.

*Testing*
**Throughput Optimized**
```
docker run --rm -it \
  -e AWS_ACCESS_KEY_ID="$AWS_ACCESS_KEY_ID" \
  -e AWS_SECRET_ACCESS_KEY="$AWS_SECRET_ACCESS_KEY" \
  -e AWS_SESSION_TOKEN="$AWS_SESSION_TOKEN" \
  -e AWS_DEFAULT_REGION="us-west-2" \
  -e AWS_PAGER="" \
  -v "${HOME}/.kube:/root/.kube" \
  -v "${PWD}:/workspace" \
  -w /workspace \
  aws-k8s-tester/kubetest2:nvidia-inference-1.31 \
    kubetest2 eksapi \
      --region=us-west-2 \
      --static-cluster-name=kubetest2-eksapi-fc73becd-27bb-4db9-b37b-c144aaaa7a53 \
      --kubeconfig=/root/.kube/config \
      --emit-metrics \
      --test=exec \
      -- \
      e2e-nvidia-inference \
      --test.timeout=60m \
      --test.v \
      --test.run=TestBertInference \
      --bertInferenceImage=031349313576.dkr.ecr.us-west-2.amazonaws.com/aws-k8s-tester/nvidia-inference:latest \
      --inferenceMode=throughput \
      --gpuRequested=1

I1231 22:43:21.585088      14 app.go:61] The files in RunDir shall not be part of Artifacts
I1231 22:43:21.585188      14 app.go:62] pass rundir-in-artifacts flag True for RunDir to be part of Artifacts
I1231 22:43:21.585204      14 app.go:64] RunDir for this run: "/workspace/_rundir/fa33654d-5081-49d5-ac08-3e14062a7268"
I1231 22:43:21.588284      14 app.go:136] ID for this run: "fa33654d-5081-49d5-ac08-3e14062a7268"
2024/12/31 22:43:21 [INFO] Applying NVIDIA device plugin.
2024/12/31 22:43:27 [INFO] NVIDIA device plugin is ready.
2024/12/31 22:43:27 [INFO] Validating cluster has at least 1 GPU(s).
2024/12/31 22:43:27 [INFO] Node ip-192-168-183-31.us-west-2.compute.internal (type: g6.8xlarge) reports 1 GPU(s).
2024/12/31 22:43:27 [INFO] Node ip-192-168-183-31.us-west-2.compute.internal (type: g6.8xlarge) meets the request of 1 GPU(s).
  torch.utils._pytree._register_pytree_node(
[2024-12-31 22:43:30,759] [INFO] [BERTInference] [INFO] Found 1 GPU(s). GPU is available.
[2024-12-31 22:43:30,759] [INFO] [BERTInference] [INFO] Running inference in throughput mode with batch size 8.
/usr/local/lib/python3.10/site-packages/huggingface_hub/file_download.py:795: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
[2024-12-31 22:43:34,922] [INFO] [BERTInference] [INFO] Adjusting num_samples from 100 to 96 to ensure full batches.
[2024-12-31 22:43:35,393] [INFO] [BERTInference] [BERT_INFERENCE_METRICS] mode=throughput avg_time_per_batch=0.016054 throughput_samples_per_sec=498.317114
2024/12/31 22:43:42 [INFO] Completed log stream for pod bert-inference-hwx49.
2024/12/31 22:43:42 [INFO] Cleaning up BERT inference job resources...
2024/12/31 22:43:42 [INFO] BERT inference job resources cleaned up.
--- PASS: TestBertInference (15.09s)
    --- PASS: TestBertInference/bert-inference (15.09s)
        --- PASS: TestBertInference/bert-inference/BERT_inference_Job_succeeds (15.04s)
PASS
2024/12/31 22:43:42 [INFO] Cleaning up NVIDIA device plugin.
2024/12/31 22:43:42 [INFO] Device plugin cleanup complete.
2024/12/31 22:43:42 [INFO] Test environment finished with exit code 0
I1231 22:43:42.596017      14 cloudwatch.go:79] emitted 1 metrics to namespace: kubetest2/eksapi
```

**Latency Optimized**
```
docker run --rm -it \
  -e AWS_ACCESS_KEY_ID="$AWS_ACCESS_KEY_ID" \
  -e AWS_SECRET_ACCESS_KEY="$AWS_SECRET_ACCESS_KEY" \
  -e AWS_SESSION_TOKEN="$AWS_SESSION_TOKEN" \
  -e AWS_DEFAULT_REGION="us-west-2" \
  -e AWS_PAGER="" \
  -v "${HOME}/.kube:/root/.kube" \
  -v "${PWD}:/workspace" \
  -w /workspace \
  aws-k8s-tester/kubetest2:nvidia-inference-1.31 \
    kubetest2 eksapi \
      --region=us-west-2 \
      --static-cluster-name=kubetest2-eksapi-fc73becd-27bb-4db9-b37b-c144aaaa7a53 \
      --kubeconfig=/root/.kube/config \
      --emit-metrics \
      --test=exec \
      -- \
      e2e-nvidia-inference \
      --test.timeout=60m \
      --test.v \
      --test.run=TestBertInference \
      --bertInferenceImage=031349313576.dkr.ecr.us-west-2.amazonaws.com/aws-k8s-tester/nvidia-inference:latest \
      --inferenceMode=latency \
      --gpuRequested=1

I1231 22:44:11.912253      15 app.go:61] The files in RunDir shall not be part of Artifacts
I1231 22:44:11.912372      15 app.go:62] pass rundir-in-artifacts flag True for RunDir to be part of Artifacts
I1231 22:44:11.912403      15 app.go:64] RunDir for this run: "/workspace/_rundir/03b0d658-68da-4857-a5e9-c33a89f7bb18"
I1231 22:44:11.915304      15 app.go:136] ID for this run: "03b0d658-68da-4857-a5e9-c33a89f7bb18"
2024/12/31 22:44:11 [INFO] Applying NVIDIA device plugin.
2024/12/31 22:44:17 [INFO] NVIDIA device plugin is ready.
2024/12/31 22:44:17 [INFO] Validating cluster has at least 1 GPU(s).
2024/12/31 22:44:17 [INFO] Node ip-192-168-183-31.us-west-2.compute.internal (type: g6.8xlarge) reports 1 GPU(s).
2024/12/31 22:44:17 [INFO] Node ip-192-168-183-31.us-west-2.compute.internal (type: g6.8xlarge) meets the request of 1 GPU(s).
2024/12/31 22:44:17 [INFO] Node ip-192-168-42-35.us-west-2.compute.internal (type: g6.8xlarge) reports 1 GPU(s).
2024/12/31 22:44:17 [INFO] Node ip-192-168-42-35.us-west-2.compute.internal (type: g6.8xlarge) meets the request of 1 GPU(s).
2024/12/31 22:44:17 [INFO] GPU capacity check passed.
=== RUN   TestBertInference
=== RUN   TestBertInference/bert-inference
2024/12/31 22:44:17 [INFO] Rendering BERT inference manifest...
2024/12/31 22:44:17 [INFO] Applying BERT inference manifest...
2024/12/31 22:44:17 [INFO] BERT inference manifest applied successfully.
=== RUN   TestBertInference/bert-inference/BERT_inference_Job_succeeds
2024/12/31 22:44:17 [INFO] Checking BERT inference job completion...
2024/12/31 22:44:32 [INFO] BERT inference job succeeded. Gathering logs...
2024/12/31 22:44:32 [INFO] BERT inference job completed in 15.007737592s
2024/12/31 22:44:32 [INFO] Pod bert-inference-b7ctq is running on node ip-192-168-183-31.us-west-2.compute.internal
2024/12/31 22:44:32 [INFO] Retrieving logs from pod bert-inference-b7ctq...
2024/12/31 22:44:32 [INFO] Logs from Pod bert-inference-b7ctq:
/usr/local/lib/python3.10/site-packages/transformers/utils/generic.py:311: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
[2024-12-31 22:44:20,935] [INFO] [BERTInference] [INFO] Running inference in latency mode with batch size 1.
/usr/local/lib/python3.10/site-packages/huggingface_hub/file_download.py:795: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
[2024-12-31 22:44:26,262] [INFO] [BERTInference] [BERT_INFERENCE_METRICS] mode=latency avg_time_per_batch=0.008406 throughput_samples_per_sec=118.956908
2024/12/31 22:44:32 [INFO] Completed log stream for pod bert-inference-b7ctq.
2024/12/31 22:44:32 [INFO] Cleaning up BERT inference job resources...
2024/12/31 22:44:32 [INFO] BERT inference job resources cleaned up.
2024/12/31 22:44:32 [INFO] Cleaning up NVIDIA device plugin.
--- PASS: TestBertInference (15.07s)
    --- PASS: TestBertInference/bert-inference (15.07s)
        --- PASS: TestBertInference/bert-inference/BERT_inference_Job_succeeds (15.03s)
PASS
2024/12/31 22:44:32 [INFO] Device plugin cleanup complete.
2024/12/31 22:44:32 [INFO] Test environment finished with exit code 0
I1231 22:44:32.813020      15 cloudwatch.go:79] emitted 1 metrics to namespace: kubetest2/eksapi
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: e2e2/test/cases/nvidia-inference/main_test.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	log.Printf("[INFO] Test environment finished with exit code %d", exitCode)
+	os.Exit(exitCode)
+}
+
+// checkGpuCapacity ensures at least one node has >= the requested number of GPUs,
+// and logs each node's instance type.
+func checkGpuCapacity(ctx context.Context, config *envconf.Config) (context.Context, error) {
+	log.Printf("[INFO] Validating cluster has at least %d GPU(s).", *gpuRequested)
+
+	cs, err := kubernetes.NewForConfig(config.Client().RESTConfig())
```
Comment (cartermckinnon) [inline e2e2/test/cases/nvidia-inference/main_test.go:96]: `config.Client()` is already a k8s API client, right? why do we need a new one?

File: e2e2/test/images/nvidia-inference/infer.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        next_sentence_labels
+    )
+
+
+def run_inference(model, tokenizer, batch_size, mode):
+    """
+    Run a dummy BERT inference workload using the given model and tokenizer.
+    Calculates average time per batch and throughput.
+    """
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (cartermckinnon) [inline e2e2/test/images/nvidia-inference/infer.py:89]: You check in main that cuda is available, so why have the fallback to cpu here?

File: e2e2/test/images/nvidia-inference/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+RUN wget -q https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tgz \
+ && tar -xzf Python-$PYTHON_VERSION.tgz \
+ && cd Python-$PYTHON_VERSION \
+ && ./configure --enable-shared --prefix=/usr/local \
+ && make -j"$(nproc)" \
+ && make install \
+ && cd .. && rm -rf Python-$PYTHON_VERSION* \
+ && ln -s /usr/local/bin/pip3 /usr/bin/pip \
+ && ln -s /usr/local/bin/$PYTHON /usr/local/bin/python \
+ && ${PIP} --no-cache-dir install --upgrade pip setuptools
```
Comment (ndbaker1) [inline e2e2/test/images/nvidia-inference/Dockerfile:70]: the `ARG PIP=pip3` seems unnecessary if its just used here

File: e2e2/test/images/nvidia-inference/infer.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        "Hiking in the mountains offers both challenge and relaxation.",
+        "Robotics and automation are revolutionizing many industries.",
+    ]
+
+    sentences_a = []
+    sentences_b = []
+    nsp_labels = []  # 1 = next, 0 = random
+
+    for _ in range(num_samples):
+        idx_a = random.randint(0, len(sample_sentences) - 1)
```
Comment (ndbaker1) [inline e2e2/test/images/nvidia-inference/infer.py:54]: a side thought i've had about these tests, should we seed the PRNG to have deterministic runs?

File: e2e2/test/images/nvidia-inference/infer.py (hunk 2 above)
Comment (ndbaker1) [reply e2e2/test/images/nvidia-inference/infer.py:89]: +1 maybe just pass `device` from main to skip the late validation

Comment (mattcjo) [reply e2e2/test/images/nvidia-inference/infer.py:89]: Good catch. Just something that carried over when developing the script. I'll have a revision taking care of this.

File: e2e2/test/images/nvidia-inference/infer.py (hunk 4 above)
Comment (mattcjo) [reply e2e2/test/images/nvidia-inference/infer.py:54]: Yes, we absolutely should be. Great catch, will set a `seed` in next revision.

File: e2e2/test/cases/nvidia-inference/main_test.go (hunk 1 above)
Comment (mattcjo) [reply e2e2/test/cases/nvidia-inference/main_test.go:96]: We need to keep using the direct Kubernetes client because the e2e framework in this version does not support cluster-scoped resource queries (like listing node objects). `config.Client().Resources(`) is namespace-based, so we need to rely on `kubernetes.NewForConfig()` to read cluster-scoped node data and validate GPU capacity.

**Review by ndbaker1 — APPROVED** [review]
